drivers/dstate.c

```diff
@@ -1002,3 +1002,163 @@ void device_alarm_commit(const int device_number)
+
+	if (!(*inited_phaseinfo) || may_reevaluate) {
+		const char *v1,  *v2,  *v3,  *v0,
+		           *v1n, *v2n, *v3n,
+		           *v12, *v23, *v31,
+		           *c1,  *c2,  *c3,  *c0;
+		char buf[80]; /* For concatenation of "xput_prefix" with items we want to query */
+		if (!xput_prefix) {
+			upsdebugx(0, "set_phasecount(): Bad xput_prefix was passed: it is NULL - function skipped");
+			return -1;
+		}
+
+		size_t xput_prefix_len = strlen(xput_prefix);
+		if (xput_prefix_len < 1) {
+			upsdebugx(0, "set_phasecount(): Bad xput_prefix was passed: it is empty - function skipped");
+			return -1;
+		}
+
+		int bufrw_max = sizeof(buf) - xput_prefix_len;
+		if (bufrw_max <= 15) {
+			/* We need to append max ~13 chars per below, so far */
+			upsdebugx(0, "set_phasecount(): Bad xput_prefix was passed: it is too long - function skipped");
+			return -1;
+		}
+		strncpy(buf, xput_prefix, sizeof(buf));
+		char *bufrw_ptr = buf + xput_prefix_len ;
+
+		/* We either have defined and non-zero (numeric) values below, or NULLs */
+		strncpy(bufrw_ptr, "L1.voltage", bufrw_max);
+		if ((v1 = dstate_getinfo(buf)))    { if (v1[0]  == '0') { v1  = NULL; } }
+		strncpy(bufrw_ptr, "L2.voltage", bufrw_max);
+		if ((v2 = dstate_getinfo(buf)))    { if (v2[0]  == '0') { v2  = NULL; } }
+		strncpy(bufrw_ptr, "L3.voltage", bufrw_max);
+		if ((v3 = dstate_getinfo(buf)))    { if (v3[0]  == '0') { v3  = NULL; } }
+		strncpy(bufrw_ptr, "L1-N.voltage", bufrw_max);
+		if ((v1n = dstate_getinfo(buf)))   { if (v1n[0] == '0') { v1n = NULL; } }
+		strncpy(bufrw_ptr, "L2-N.voltage", bufrw_max);
+		if ((v2n = dstate_getinfo(buf)))   { if (v2n[0] == '0') { v2n = NULL; } }
+		strncpy(bufrw_ptr, "L3-N.voltage", bufrw_max);
+		if ((v3n = dstate_getinfo(buf)))   { if (v3n[0] == '0') { v3n = NULL; } }
+		strncpy(bufrw_ptr, "L1-L2.voltage", bufrw_max);
+		if ((v12 = dstate_getinfo(buf)))   { if (v12[0] == '0') { v12 = NULL; } }
+		strncpy(bufrw_ptr, "L2-L3.voltage", bufrw_max);
+		if ((v23 = dstate_getinfo(buf)))   { if (v23[0] == '0') { v23 = NULL; } }
+		strncpy(bufrw_ptr, "L3-L1.voltage", bufrw_max);
+		if ((v31 = dstate_getinfo(buf)))   { if (v31[0] == '0') { v31 = NULL; } }
+		strncpy(bufrw_ptr, "L3.current", bufrw_max);
+		if ((c1 = dstate_getinfo(buf)))    { if (c1[0]  == '0') { c1  = NULL; } }
+		strncpy(bufrw_ptr, "L2.current", bufrw_max);
+		if ((c2 = dstate_getinfo(buf)))    { if (c2[0]  == '0') { c2  = NULL; } }
+		strncpy(bufrw_ptr, "L3.current", bufrw_max);
+		if ((c3 = dstate_getinfo(buf)))    { if (c3[0]  == '0') { c3  = NULL; } }
+		strncpy(bufrw_ptr, "voltage", bufrw_max);
+		if ((v0 = dstate_getinfo(buf)))    { if (v0[0]  == '0') { v0  = NULL; } }
+		strncpy(bufrw_ptr, "current", bufrw_max);
+		if ((c0 = dstate_getinfo(buf)))    { if (c0[0]  == '0') { c0  = NULL; } }
+
+		if ( (v1 && v2 && v3) ||
+		     (v1n && v2n && v3n) ||
+		     (c1 && (c2 || c3)) ||
+		     (c2 && (c1 || c3)) ||
+		     (c3 && (c1 || c2)) ||
```

Comment:
Yes, mathematically your version is correct and equivalent :)

But I feel mine is clearer to read. And for review (or authorship) any aid to avoid a stupid mistake and not outsmart oneself or others is useful :)